src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs

```diff
@@ -173,92 +201,115 @@ public CheckResult Check(string expressionText, RecordType parameterType, Parser
+
+            CheckWorker(check);
+            return check;
         }
 
-        public CheckResult Check(
-            ParseResult parse,
-            ParserOptions options = null,
-            ReadOnlySymbolTable symbolTable = null)
+        // Apply a standard set of operations on the CheckResult.
+        // If callers want more granularity, they can create the CheckResult themselves. 
+        private void CheckWorker(CheckResult check)
         {
-            var bindingConfig = new BindingConfig(allowsSideEffects: options?.AllowsSideEffects == true);
+            check.ApplyBindingInternal();
+            check.ApplyErrors();
+            check.ApplyDependencyAnalysis();
+        }
 
-            return CheckInternal(parse, bindingConfig, symbolTable);
+        // Called after check result, can inject additional errors or constraints. 
+        protected virtual IEnumerable<ExpressionError> PostCheck(CheckResult check)
+        {
+            return new ExpressionError[0];
         }
 
-        private CheckResult CheckInternal(
-            ParseResult parse,
-            BindingConfig bindingConfig = null,
-            ReadOnlySymbolTable symbolTable = null)
+        internal IEnumerable<ExpressionError> InvokePostCheck(CheckResult check)
+        {
+            return this.PostCheck(check);
+        }
+
+        // Setting rule sope which will get passed into Binder. 
+        // Prefer to avoid this hook and use SymbolTables instead. 
+        private protected virtual RecordType GetRuleScope()
+        {
+            return null;
+        }
+
+        private BindingConfig GetDefaultBindingConfig()
+        {
+            var ruleScope = this.GetRuleScope();
+            bool useThisRecordForRuleScope = ruleScope != null;
+
+            var bindingConfig = BindingConfig.Default;
+
+            if (useThisRecordForRuleScope)
+            {
+                bindingConfig = new BindingConfig(bindingConfig.AllowsSideEffects, true);
+            }
+
+            return bindingConfig;
+        }
+
+        // Called by CheckResult.ApplyBinding to compute the binding. 
+        internal (TexlBinding, ReadOnlySymbolTable) ComputeBinding(CheckResult result)
         {
-            var startHash = symbolTable?.VersionHash;
+            var parse = result.ApplyParse();
+
+            ReadOnlySymbolTable symbolTable = result.Parameters;
 
             // Ok to continue with binding even if there are parse errors. 
             // We can still use that for intellisense.             
             var resolver = CreateResolverInternal(out var combinedSymbols, symbolTable);
-            
+
             var glue = CreateBinderGlue();
 
+            var ruleScope = this.GetRuleScope();
+
+            // Canvas apps uses rule scope for lots of cases. 
+            // But in general, we should only use rule scope for 'ThisRecord' binding. 
+            // Anything else should be accomplished with SymbolTables.
```

Comment:
Shall we work with them to get this changed?

Reply:
Ideally, but it will be very challenging to get them to prioritize it. 
It's well outside of scope for this PR.